File: tests/visualization/barLC.py
import subprocess
import numpy as np
import matplotlib.pyplot as plt
import numpy.fft as fft
import tools
import style

# Get host and user name.
host_name = subprocess.getoutput("hostname")
user_name = subprocess.getoutput("whoami")

if "levante" in host_name:
  # Levante cluster
  data_path = "/scratch/b/" + user_name + "/PF/runs"
  reference_path = "/scratch/b/" + user_name + "/PF/pinc/reference"

elif "login" in host_name:
  # Goethe cluster
  data_path = "/scratch/atmodynamics/" + user_name + "/PF/runs"
  reference_path = "/scratch/atmodynamics/" + user_name + "/PF/pinc/reference"

else:
  # Local machine
  data_path = "../"
  reference_path = "../"

print("data_path =", data_path)
print("reference_path =", reference_path)

# Import data.
data = tools.ModelOutput(data_path + 'barLC/')
reference = tools.ModelOutput(reference_path + 'barLC/')

# Set slices.
it = - 1
iz = 2
iy = slice(42, None)

print(" ".join(("Time:", str(data.variables['t'][it]), "s")))

# Set grid.
xx = data.variables['x'][:] * 0.001
yy = data.variables['y'][iy] * 0.001
dx = np.abs(xx[1] - xx[0]) * 1000.0
dy = np.abs(yy[1] - yy[0]) * 1000.0

# Loop over data and reference.
for data_set in (reference, data):

  # Set fields of interest.
  uu = data_set.groups['atmvar'].variables['u'][it, iz, iy]
  vv = data_set.groups['atmvar'].variables['v'][it, iz, iy]
  thetap = data_set.groups['atmvar'].variables['thetap'][it, iz, iy]

  # Compute divergence.
  divergence = np.zeros_like(uu)
  divergence[1:, 1:] = (uu[1:, 1:] - uu[1:, :(- 1)]) / dx + (vv[1:, 1:] - \
      vv[:(- 1), 1:]) / dy

  # Apply Fourier filter.
  urossby = np.zeros_like(uu)
  ugravity = np.zeros_like(uu)
  sigma = divergence
  sigmatilde = fft.fft2(sigma)
  kk = fft.fftfreq(sigma.shape[1], d = dx)
  ll = fft.fftfreq(sigma.shape[0], d = dy)
  urossbytilde = sigmatilde.copy()
  urossbytilde[np.abs(ll) > 1.0e-6, :] = 0.0
  urossbytilde[:, np.abs(kk) > 1.0e-6] = 0.0
  urossby = fft.ifft2(urossbytilde).real
  ugravity = sigma - urossby

  # Compute difference of relevant fields.
  if data_set == reference:
    deltaugravity = ugravity.copy()
    deltatheta = thetap.copy()
  elif data_set == data:
    deltaugravity = ugravity - deltaugravity
    deltatheta = thetap - deltatheta

# Create plot.
(levels, colormap) = style.symmetric_contours(ugravity.min(), ugravity.max())
fig, axes = plt.subplots()
plot = axes.contourf(xx, yy, ugravity, levels, cmap = colormap)
axes.contour(xx, yy, thetap, linewidths = 1.0, colors = "black")
axes.set_xlabel(r"$x \, \left[\mathrm{km}\right]$")
axes.set_ylabel(r"$y \, \left[\mathrm{km}\right]$")
fig.colorbar(plot, label = r"$\boldsymbol{\nabla}_z \cdot" r"\boldsymbol{u}" \
    r"\, \left[\mathrm{s^{- 1}}\right]$")
fig.savefig("".join((data_path, "/results/barLC.pdf")))
fig.savefig("".join((data_path, "/results/barLC.png")), dpi = 500)

# Create difference plot.
if data_path != reference_path:
  (levels, colormap) = style.symmetric_contours(deltaugravity.min(), \
      deltaugravity.max())
  fig, axes = plt.subplots()
  plot = axes.contourf(xx, yy, deltaugravity, levels, cmap = colormap)
  axes.contour(xx, yy, deltatheta, linewidths = 1.0, colors = "black")
  axes.set_xlabel(r"$x \, \left[\mathrm{km}\right]$")
  axes.set_ylabel(r"$y \, \left[\mathrm{km}\right]$")
  fig.colorbar(plot, label = r"$\Delta \boldsymbol{\nabla}_z \cdot" \
      r"\boldsymbol{u} \, \left[\mathrm{s^{- 1}}\right]$")
  fig.savefig("".join((data_path, "/results/barLC_difference.pdf")))
  fig.savefig("".join((data_path, "/results/barLC_difference.png")), \
      dpi = 500)

File: tests/visualization/barLC.jl
include("style.jl")
include("tools.jl")

using FFTW
using LaTeXStrings

# Set paths.
data_path = "../"
reference_path = "../"

println("data_path =", data_path)
println("reference_path =", reference_path)

# Import data.
data = ModelOutput(data_path * "barLC")
reference = ModelOutput(reference_path * "barLC")

println("Time: ", data.variables["t"][end], " s")

# Set grid.
x = data.variables["x"][:] .* 0.001
y = data.variables["y"][43:end] .* 0.001
dx = abs(x[2] - x[1]) * 1000.0
dy = abs(y[2] - y[1]) * 1000.0
x = [xi for xi in x, j in 1:size(y)[1]]
y = [yj for i in 1:size(x)[1], yj in y]

# Loop over data and reference.
ugravity = Dict()
thetap = Dict()
for (path, data_set) in zip((reference_path, data_path), (reference, data))

  # Set fields of interest.
  u = data_set.groups["atmvar"].variables["u"][:, 43:end, 3, end]
  v = data_set.groups["atmvar"].variables["v"][:, 43:end, 3, end]
  thetap[path] =
    data_set.groups["atmvar"].variables["thetap"][:, 43:end, 3, end]

  # Compute divergence.
  divergence = zeros(size(u))
  for j in 2:(size(u)[2] - 1), i in 2:(size(u)[1] - 1)
    divergence[i, j] =
      0.5 *
      ((u[i + 1, j] - u[i - 1, j]) / dx + (v[i, j + 1] - v[i, j - 1]) / dy)
  end

  # Apply Fourier filter.
  sigma = divergence
  sigmatilde = fft(sigma)
  k = fftfreq(size(sigma)[1], 1.0 / dx)
  l = fftfreq(size(sigma)[2], 1.0 / dy)
  urossbytilde = deepcopy(sigmatilde)
  urossbytilde[:, abs.(l) .> 1.0e-6] .= 0.0
  urossbytilde[abs.(k) .> 1.0e-6, :] .= 0.0
  urossby = real.(ifft(urossbytilde))
  ugravity[path] = sigma .- urossby
end

# Compute difference of relevant fields.
deltaugravity = ugravity[data_path] .- ugravity[reference_path]
deltathetap = thetap[data_path] .- thetap[reference_path]
ugravity = ugravity[data_path]
thetap = thetap[data_path]

# Create plot.
(levels, colormap) = symmetric_contours(minimum(ugravity), maximum(ugravity))
contours = contourf(x, y, ugravity; levels = levels, cmap = colormap)
contour(x, y, thetap; linewidths = 1.0, colors = "black")
xlabel(L"x\,\left[\mathrm{km}\right]")
ylabel(L"y\,\left[\mathrm{km}\right]")
colorbar(
  contours;
  label = L"\boldsymbol{\nabla}_z\cdot\boldsymbol{u}\," *
          L"\left[\mathrm{s^{-1}}\right]",
)
savefig(data_path * "/results/barLC.png")

# Create difference plot.
if data_path != reference_path
  (levels, colormap) =
    symmetric_contours(minimum(deltaugravity), maximum(deltaugravity))
  clf()
  contours = contourf(x, y, deltaugravity; levels = levels, cmap = colormap)
  contour(x, y, deltathetap; linewidths = 1.0, colors = "black")
  xlabel(L"x\,\left[\mathrm{km}\right]")
  ylabel(L"y\,\left[\mathrm{km}\right]")
  colorbar(
    contours;
    label = L"\Delta\boldsymbol{\nabla}_z\cdot\boldsymbol{u}" *
            L"\,\left[\mathrm{s^{-1}}\right]",
  )
  savefig(data_path * "/results/barLC_difference.png")
end
